emoji/src/main/java/com/vanniktech/emoji/RecentVariantManager.java

```diff
@@ -0,0 +1,145 @@
+package com.vanniktech.emoji;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.support.annotation.NonNull;
+import com.vanniktech.emoji.emoji.Emoji;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.StringTokenizer;
+
+final class RecentVariantManager { // TODO: test it
+  private static final RecentVariantManager INSTANCE = new RecentVariantManager();
+
+  public static RecentVariantManager getInstance() {
+    return INSTANCE;
+  }
+
+  private static final String PREFERENCE_NAME = "emoji-recent-skintone-manager";
+  private static final String EMOJI_DELIMITER = "~";
+  private static final String RECENT_VARIANTS = "recent-skintones";
+  private static final int EMOJI_GUESS_SIZE = 5;
+
+  private List<Emoji> recentVariantsList = new ArrayList<>(0);
+
+  Map<Emoji, EmojiVariantListener> variantListeners = new HashMap<>();
+
+  Emoji getMostRecentVariant(@NonNull final Emoji desiredEmoji, final Context context) {
+    if (recentVariantsList.isEmpty()) {
+      initFromSharedPreferences(context);
+    }
+
+    final Emoji baseEmoji = desiredEmoji.getBase();
+
+    if (recentVariantsList.size() > 0) {
+      for (final Emoji emoji : recentVariantsList) {
+        if (baseEmoji.equals(emoji.getBase())) {
+          return emoji;
+        }
+      }
+    }
+
+    return desiredEmoji;
+  }
+
+  void addRecentVariant(@NonNull final Emoji newVariant) {
+    final Emoji newVariantBase = newVariant.getBase();
+
+    // TODO: test that going from a skin tone to a non skin tone works.
+    //if (newVariantBase.equals(newVariant)) { // Has no skin-tone.
+    //  return;
+    //}
+
+    // TODO: sometimes those 4 characters are not working (e.g. 270a)
+
+    for (final Emoji recentEmoji : recentVariantsList) {
+      if (recentEmoji.getBase().equals(newVariantBase)) {
+        if (recentEmoji.equals(newVariant)) { // Same skin-tone was used.
+          return;
+        } else {
+          recentVariantsList.remove(recentEmoji);
```

Comment:
This is potentially slower than possible.

`remove` needs to find the `recentEmoji` again, by iteration the list, even though we have already found it.
When using `fori` in which we memorize the index, we can avoid that.